File: leetcode/121.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
 public:
  int maxProfit(vector<int>& p) {
    int ret = 0;
    int prevMax = 0;
    for (auto it = p.rbegin(); it != p.rend(); it++) {
      ret = max(ret, prevMax - *it);
      prevMax = max(prevMax, *it);
    }
    return ret;
  }
};

File: leetcode/121.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(p) {
  let ret = 0;
  let prevMax = 0;
  for (let i = p.length - 1; i >= 0; i--) {
    ret = Math.max(ret, prevMax - p[i]);
    prevMax = Math.max(prevMax, p[i]);
  }
  return ret;
};
